### PR TITLE
COMMON:

### PR DESCRIPTION
**Pull request by manolama**
- Add rehashing hack to the LongLongHashTable wherein we track the average
  number of scans (which will increase after deleting some data) and rehash
  when it hits a given threshold.
  TODO - configure that threshold.

### CODE REVIEW

File: common/src/main/java/net/opentsdb/collections/LongLongHashTable.java [hunk 1]
```diff
@@ -43,17 +61,69 @@
...
+  private int prime;
+  private long scans;
+  private long opCount;
+  private double scanRehashThreshold;
+  private float growthRate;
+  private float growthDecay;
   private DirectByteArray table;
 
+  /**
+   * Default ctor with a 0.75 load, 0.75 growth factor, 0.75 growth decline and
```
Comment (smrutilal2) [inline common/src/main/java/net/opentsdb/collections/LongLongHashTable.java:74]: typo, default growth decay is `0.25`

File: common/src/main/java/net/opentsdb/collections/LongLongHashTable.java [hunk 2]
```diff
@@ -78,16 +154,21 @@ public void close() {
...
 
   private void resize() {
-
+    long start = System.nanoTime();
+    int oldLength = arrayLength;
     int oldCap = slots;
     int newCap = oldCap;
     if (table != null) { // growing the table;
-      newCap = oldCap << 1; // double the size;
+      newCap = (int) (oldCap + (oldCap * growthRate));
```
Comment (smrutilal2) [inline common/src/main/java/net/opentsdb/collections/LongLongHashTable.java:162]: Now, after the table is grown, the new `size` may not be a prime. In such case the modulo hash `(key % size)` can cause collisions. That might be okay for now, but something to keep in mind.

File: common/src/main/java/net/opentsdb/collections/LongLongHashTable.java [hunk 3]
```diff
@@ -112,9 +195,45 @@ private void resize() {
...
+    }
+    prime = PRIMES_FOR_HASHING[primeIndex];
+    scans = 0;
+    opCount = 0;
+
+    long oldAddress = table.init(arrayLength, false);
+    DirectByteArray oldTable = new DirectByteArray(oldAddress, false, slots * slotSz);
+    this.address = table.getAddress();
+    this.sz = 0;
+    if (oldTable != null) {
```
Comment (smrutilal2) [inline common/src/main/java/net/opentsdb/collections/LongLongHashTable.java:217]: redundant null check

**Review by smrutilal2 — COMMENTED** [review]
Overall looks good. Just couple of typo and redundant null check

File: common/src/main/java/net/opentsdb/collections/LongLongHashTable.java (hunk 2 above)
Comment (manolama) [reply common/src/main/java/net/opentsdb/collections/LongLongHashTable.java:162]: True, we have to keep an eye on that.

**Review by smrutilal2 — APPROVED** [review]

**Pull request merged**